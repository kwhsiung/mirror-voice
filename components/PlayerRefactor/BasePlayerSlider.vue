<template>
  <VueSlider
    v-model="value"
    class="slider"
    :style="{ padding: '0' }"
    :rail-style="{ background }"
    v-bind="options"
  />
</template>

<script>
import VueSlider from 'vue-slider-component/dist-css/vue-slider-component.umd.min.js'

export default {
  components: {
    VueSlider
  },
  props: {
    direction: {
      type: String,
      required: true,
      validator(value) {
        return ['horizontal', 'vertical'].includes(value)
      }
    },
    valueBuffered: {
      type: Number,
      default: 0,
      require: true,
      validator(value) {
        return value >= 0 && value <= 1
      }
    },
    valueCurrent: {
      type: Number,
      default: 0,
      require: true,
      validator(value) {
        return value >= 0 && value <= 1
      }
    }
  },
  computed: {
    value: {
      get() {
        return this.valueCurrent
      },
      set(value) {
        this.$emit('update:valueCurrent', value)
      }
    },
    options() {
      return {
        interval: 0.001,
        direction: this.direction === 'horizontal' ? 'ltr' : 'btt',
        width: this.direction === 'horizontal' ? 'auto' : 4,
        height: this.direction === 'horizontal' ? 4 : 126,
        min: 0,
        max: 1,
        tooltip: 'none',
        dotSize: 12,
        processStyle: {
          backgroundColor: '#d84939'
        }
      }
    },
    background() {
      const buffered = this.valueBuffered * 100
      return `linear-gradient(90deg, #7d7d7d ${buffered}%, black ${buffered}%)`
    }
  }
}
</script>

<style lang="stylus" scoped>
.slider
  cursor pointer
  position relative
  & >>> .vue-slider-dot
    background-color #eeeeee
    border-radius 100%
    position absolute
  & >>> .vue-slider-rail
    height 100%
</style>
