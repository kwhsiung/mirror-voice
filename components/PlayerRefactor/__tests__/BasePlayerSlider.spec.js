import { shallowMount } from '@vue/test-utils'
import VueSlider from 'vue-slider-component'
import BasePlayerSlider from '../BasePlayerSlider.vue'

describe('BasePlayerSlider', () => {
  test('create different vueSliderComponentOption object by different "direction" props', () => {
    let wrapper = shallowMount(BasePlayerSlider, {
      propsData: {
        direction: 'horizontal'
      }
    })
    let slider = wrapper.find(VueSlider)

    expect(slider.props().direction).toBe('ltr')
    expect(slider.props().width).toBe('auto')
    expect(slider.props().height).toBe(4)

    wrapper = shallowMount(BasePlayerSlider, {
      propsData: {
        direction: 'vertical'
      }
    })
    slider = wrapper.find(VueSlider)

    expect(slider.props().direction).toBe('btt')
    expect(slider.props().width).toBe(4)
    expect(slider.props().height).toBe(126)
  })
  test('passing value received from "valueCurrent" props to slider', () => {
    const valueCurrent = 0.5
    const wrapper = shallowMount(BasePlayerSlider, {
      propsData: {
        direction: 'horizontal',
        valueCurrent
      }
    })
    const slider = wrapper.find(VueSlider)

    expect(slider.props().value).toBe(valueCurrent)
  })
  test('emit "update:valueCurrent" event while slider value change', () => {
    const valueCurrent = 0.5
    const wrapper = shallowMount(BasePlayerSlider, {
      propsData: {
        direction: 'horizontal',
        valueCurrent
      }
    })
    const slider = wrapper.find(VueSlider)

    const valueChanged = 0.6

    // wrapper.setValue() only works on text-control input
    // use "change" event to trigger v-model instead
    slider.vm.$emit('change', valueChanged)

    expect(wrapper.emitted('update:valueCurrent')[0]).toEqual([valueChanged])
  })
  test('passing proper "rail-style" props computed by "valueBuffered" props to slider', () => {
    const valueBuffered = 0.5
    const wrapper = shallowMount(BasePlayerSlider, {
      propsData: {
        direction: 'horizontal',
        valueBuffered
      }
    })
    const slider = wrapper.find(VueSlider)

    expect(slider.props().railStyle).toEqual({
      background: `linear-gradient(90deg, #7d7d7d ${valueBuffered *
        100}%, black ${valueBuffered * 100}%)`
    })
  })
})
