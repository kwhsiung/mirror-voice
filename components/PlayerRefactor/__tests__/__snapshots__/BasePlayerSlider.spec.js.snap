// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders component correctly if we provide efficient props in horizontal 1`] = `
<vueslider-stub
  class="slider"
  clickable="true"
  direction="ltr"
  dotsize="12"
  duration="0.5"
  enablecross="true"
  height="4"
  interval="0.001"
  max="1"
  min="0"
  order="true"
  process="true"
  processstyle="[object Object]"
  railstyle="[object Object]"
  style="padding: 0px;"
  tooltip="none"
  value="0.3"
  width="auto"
/>
`;

exports[`snapshot tests renders component correctly if we provide efficient props in vertical 1`] = `
<vueslider-stub
  class="slider"
  clickable="true"
  direction="btt"
  dotsize="12"
  duration="0.5"
  enablecross="true"
  height="126"
  interval="0.001"
  max="1"
  min="0"
  order="true"
  process="true"
  processstyle="[object Object]"
  railstyle="[object Object]"
  style="padding: 0px;"
  tooltip="none"
  value="0.3"
  width="4"
/>
`;
